File: app/javascript/channels/chatroom_channel.js
import consumer from './consumer'

$(window).on('load',function() {
  if(document.URL.match(/\/.+chat_rooms\/\d+/)) {
    var room_id = gon.chatroom_id;
    var chatChannel = consumer.subscriptions.create({ channel: "ChatroomChannel", room_id: room_id }, {
      connected: function() {},
      disconnected: function() {},
      received: function(data) {
        $('.messages').append(data['content']);
      },
      speak: function(message) {
        return this.perform('speak', {
          room_id: room_id,
          user_id: $('meta[name="current_user_id"]').attr('content'),
          photographer_id: $('meta[name="current_photographer_id"]').attr('content'),
          message: message
        });
      }
    });
    $(document).on('click', '#message_submit', function(event) {
      var text = $(this).prev();
      if ($.trim(text.val()).length > 1) {
        chatChannel.speak(text.val());
        text.val('');
        $('.messages').animate({scrollTop:$('.messages')[0].scrollHeight});
        return event.preventDefault();
      }
    });
    $('.messages').animate({scrollTop:$('.messages')[0].scrollHeight});
  }
});

File: app/channels/chatroom_channel.rb
class ChatroomChannel < ApplicationCable::Channel
  def subscribed
    stream_for "chatroom_#{params[:room_id]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    message = ChatMessage.create!(chatroom_id: data['room_id'], user_id: data['user_id'], photographer_id: data['photographer_id'], message: data['message'])
    ChatroomChannel.broadcast_to "chatroom_#{data['room_id']}", content: render_message(message)
  end
  
  private

  def render_message(message)
    ApplicationController.renderer.render(partial: 'shared/chat_message', locals: { chat_message: message })
  end

end
